app/rust/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+
+    unsafe {
+        let output_slice = &mut *output;
+        output_slice.copy_from_slice(&result.to_bytes());
+    }
+
+    ParserError::ParserOk
+}
+
+#[inline(never)]
+pub fn get_diversifiers(
+    dk: &[u8; 32],
+    start_index: &mut [u8; 11],
+    result: &mut [u8; 44],
+) {
+    // Initialize cipher
+    let key = GenericArray::from_slice(dk);
+    let cipher = Aes256::new(&key);
+    let mut scratch = [0; 12];
+
+    let mut ff1 = BinaryFF1::new(&cipher, 11, &[], &mut scratch).unwrap();
+
+    let mut d: [u8; 11];
+    let size = 4;
+
+    for c in 0..size {
+        d = *start_index;
+        ff1.encrypt(&mut d).unwrap();
+        result[c * 11..(c + 1) * 11].copy_from_slice(&d);
+        for k in 0..11 {
+            start_index[k] = start_index[k].wrapping_add(1);
+            if start_index[k] != 0 {
+                // No overflow
+                break;
+            }
+        }
+    }
+}
+
+#[inline(never)]
+fn diversifier_group_hash_light(tag: &[u8]) -> bool {
+    if tag == [0u8; 11] {
+        return false;
+    }
+    let h = Blake2sParams::new()
+        .hash_length(32)
+        .personal(KEY_DIVERSIFICATION_PERSONALIZATION)
+        .to_state()
+        .update(GH_FIRST_BLOCK)
+        .update(tag)
+        .finalize();
+
+    let u = AffinePoint::from_bytes(*h.as_array());
+    if u.is_some().unwrap_u8() == 1 {
+        let q = u.unwrap().mul_by_cofactor();
+        return q != ExtendedPoint::identity();
+    }
+
+    false
+}
+
+#[inline(never)]
+fn is_valid_diversifier(div_ptr: *const [u8; 11]) -> bool {
+    let div = unsafe { &*div_ptr };
+    diversifier_group_hash_light(div)
+}
+
+#[no_mangle]
+pub extern "C" fn get_default_diversifier(
+    dk: &[u8; 32],
+    start: &mut [u8; 11],
+    d: &mut [u8; 11],
+) -> ParserError {
+    let mut found = false;
+    let mut div_list = [0u8; DIV_SIZE * DIV_DEFAULT_LIST_LEN];
+
+    while !found {
+        get_diversifiers(dk, start, &mut div_list);
+        for i in 0..DIV_DEFAULT_LIST_LEN {
+            if !found
+                && is_valid_diversifier(
+                    &div_list[i * DIV_SIZE..(i + 1) * DIV_SIZE]
+                        .try_into()
+                        .unwrap(),
+                )
+            {
+                found = true;
+                d.copy_from_slice(&div_list[i * DIV_SIZE..(i + 1) * DIV_SIZE]);
+            }
+        }
+    }
+
+    ParserError::ParserOk
+}
+
+#[no_mangle]
+pub extern "C" fn get_default_diversifier_list(
+    dk: &[u8; 32],
+    start_index: &mut [u8; 11],
+    d_l: &mut [u8; 44],
+) -> ParserError {
+    let start = &mut *start_index;
+    let diversifier =  &mut *d_l;
+    get_diversifiers(dk,  start, diversifier);
+    ParserError::ParserOk
+}
+
+#[no_mangle]
+pub extern "C" fn get_pkd(
+    ivk_ptr: &[u8; 32],
+    d: & [u8; 11],
+    pk_d: &mut [u8; 32],
+) -> ParserError {
+
+    let h = Blake2sParams::new()
+        .hash_length(32)
+        .personal(KEY_DIVERSIFICATION_PERSONALIZATION)
+        .to_state()
```

Comment:
same as above, we can move Blake2s hash to C and let the math operation from 192~196 to Rust